urld5c972ef81_noxx90_jflam_tgz-pJ8-fr_noxx90_jflam_BenchmarkJ8/src/fr/noxx90/jflam/renderer/impl/ThreadPooledHistogramRenderer.java

```diff
@@ -0,0 +1,49 @@
+package fr.noxx90.jflam.renderer.impl;
+
+import static com.esotericsoftware.minlog.Log.info;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import fr.noxx90.jflam.model.Flame;
+import fr.noxx90.jflam.model.Histogram;
+import fr.noxx90.jflam.model.impl.InMemoryHistogram;
+
+public class ThreadPooledHistogramRenderer extends DefaultHistogramRenderer
+{
+	protected ThreadPoolExecutor executor;
+	
+	public ThreadPooledHistogramRenderer(int maxPoolSize) {
+		BlockingQueue<Runnable> queue = new ArrayBlockingQueue<Runnable>(100);
+		executor = new ThreadPoolExecutor(maxPoolSize, maxPoolSize, 5, TimeUnit.SECONDS, queue);
+	}
+	
+	@Override
+	public Histogram render(Flame flame, int width, int height, int quality, int sampling) {
+		info("DefaultHistogramRenderer", "first creating histogram with : width=" + width + ", height=" + height + ", quality=" + quality + ", sampling=" + sampling);
+		initializeRangeMap(flame);
+		
+		Histogram histogram = new InMemoryHistogram(width, height);
+		initializeProjection(histogram);
+		
+		int points = Math.round((float) quality / 100 * width * height);
+		info("DefaultHistogramRenderer", "number of random points : " + points);
+		
+		for(int i = 0;i < 100;i++) {
+			RendererTask task = new RendererTask(rangeMap, weightSum, ratio, marginTop, marginLeft, histogram, points/100, width, height);
+			executor.execute(task);
+		}
+		
+		try {
+			executor.shutdown();
+			executor.awaitTermination(1, TimeUnit.DAYS);
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)